packages/ts/react-signals/src/ValueSignal.ts

```diff
@@ -86,20 +129,25 @@ export class ValueSignal<T> extends FullStackSignal<T> {
     const record = this.#pendingRequests.get(event.id);
     if (record) {
       this.#pendingRequests.delete(event.id);
-    }
 
-    if (!event.accepted && record) {
-      if (!record.canceled) {
-        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      if (!(event.accepted || record.canceled)) {
         this.update(record.callback);
       }
     }
 
     if (event.accepted || isSnapshotStateEvent<T>(event)) {
-      if (record) {
-        record.waiter.resolve();
-      }
       this.#applyAcceptedEvent(event);
+      // `then` callbacks can be associated to the record or the event
+      // it depends on the operation that was performed
+      [record?.id, event.id].filter(Boolean).forEach((id) => this.runThenCallback(id!));
+    }
+  }
+
+  protected runThenCallback(eventId: string): void {
```

Comment:
Could the function signature be modified to resemble other `protected` methods in [FullStackSignal.ts](https://github.com/vaadin/hilla/blob/main/packages/ts/react-signals/src/FullStackSignal.ts#L212)?